src/typestate-programming/zero-cost-abstractions.md

```diff
@@ -0,0 +1,78 @@
+# Zero Cost Abstractions
+
+## "no runtime cost"?
+
+```rust
+use core::mem::size_of;
+
+let _ = size_of::<GpioPin>();     // == 0
+let _ = size_of::<InputGpio>();   // == 0
+let _ = size_of::<OutputGpio>();  // == 0
+let _ = size_of::<()>();          // == 0
+```
+
+## Zero Sized Types
+
+```rust
+struct GpioPin;
+```
+
+> acts real at compile time
+>
+> doesn't exist in the binary
+>
+> no RAM, no CPU, no space
+>
+> Evaporates at compile time
+
+## What if our `OutputGpio` has multiple modes?
+
+> (it does)
+
+---
+
+```rust
+pub struct PushPull;  // good for general usage
+pub struct OpenDrain; // better for high power LEDs
```

Comment:
In a MCU the main usage of OpenDrain GPIOs is communication with several devices on one bus. The internal MOSFETs in MCUs typically do not have a very high power rating, often not even higher than what a PP configuration can sink or source and also the maximum power is usually limited by the general sinking capabilities of the MCU which usually is in the order of a few hundred mA.